src/plot_api/plot_api.js

```diff
@@ -2362,6 +2362,193 @@ exports._guiRestyle = guiEdit(restyle);
+    {pattern: /colorbar\.(x|y)$/, attr: 'editrevision'}
+];
+
+function findUIPattern(key, patternSpecs) {
+    for(var i = 0; i < patternSpecs.length; i++) {
+        var spec = patternSpecs[i];
+        var match = key.match(spec.pattern);
+        if(match) {
+            return {head: match[1], attr: spec.attr, autofill: spec.autofill};
+        }
+    }
+}
+
+// We're finding the new uirevision before supplyDefaults, so do the
+// inheritance manually. Note that only `undefined` inherits - other
+// falsy values are returned.
+function getNewRev(revAttr, container) {
+    var newRev = nestedProperty(container, revAttr).get();
+    if(newRev !== undefined) return newRev;
+
+    var parts = revAttr.split('.');
+    parts.pop();
+    while(parts.length > 1) {
+        parts.pop();
+        newRev = nestedProperty(container, parts.join('.') + '.uirevision').get();
+        if(newRev !== undefined) return newRev;
+    }
+
+    return container.uirevision;
+}
+
+function getFullTraceIndexFromUid(uid, fullData) {
+    for(var i = 0; i < fullData.length; i++) {
+        if(fullData[i]._fullInput.uid === uid) return i;
+    }
+    return -1;
+}
+
+function getTraceIndexFromUid(uid, data, tracei) {
+    for(var i = 0; i < data.length; i++) {
+        if(data[i].uid === uid) return i;
+    }
+    // fall back on trace order, but only if user didn't provide a uid for that trace
+    return data[tracei].uid ? -1 : tracei;
+}
+
+function applyUIRevisions(data, layout, oldFullData, oldFullLayout) {
+    var layoutPreGUI = oldFullLayout._preGUI;
+    var key, revAttr, oldRev, newRev, match, preGUIVal, newNP, newVal;
+    for(key in layoutPreGUI) {
+        match = findUIPattern(key, layoutUIControlPatterns);
+        if(match) {
+            revAttr = match.attr || (match.head + '.uirevision');
+            oldRev = nestedProperty(oldFullLayout, revAttr).get();
+            newRev = oldRev && getNewRev(revAttr, layout);
+            if(newRev && (newRev === oldRev)) {
+                preGUIVal = layoutPreGUI[key];
+                if(preGUIVal === null) preGUIVal = undefined;
+                newNP = nestedProperty(layout, key);
+                newVal = newNP.get();
+                // TODO: This test for undefined is to account for the case where
+                // the value was filled in automatically in gd.layout,
+                // like axis.range/autorange. In principle though, if the initial
+                // plot had a value and the new plot removed that value, we would
+                // want the removal to override the GUI edit and generate a new
+                // auto value. But that would require figuring out what value was
+                // in gd.layout *before* the auto values were filled in, and
+                // storing *that* in preGUI... oh well, for now at least I limit
+                // this to attributes that get autofilled, which AFAICT among
+                // the GUI-editable attributes is just axis.range/autorange.
```

Comment:
A fairly involved and kind of strange edge case, but I still may try to address it before we merge this:
- the initial plot has an explicit range
- the user zooms/pans to a different range
- `react` is called again, with the SAME `uirevision` but the explicit range is removed (this step seems weird to me, but it could happen...)

The correct behavior here would be to autorange, but currently we instead keep the GUI-defined range, since we can't tell that the original pre-GUI range wasn't actually something we filled in automatically during plotting. If we ever get around to getting rid of that late `gd.layout` mutation this problem will go away.

Reply:
That looks ok to me. Great tests :+1: